Careful handling when loading files

jentfoo opened the pull request:
This PR attempts to provide a common API so that the decision of when to follow symlinks is a conscious decision. Because Teleport (particularly the agent) runs in a privileged context, there is risk that following symlinks may allow information disclosure.

After review of the cases covered in this commit (and some additional cases where this API was not a natural fit), this does not appear to be a broad problem.  This PR however does fix the one known flaw described in the issue https://github.com/gravitational/teleport-private/issues/1009

In many cases where this API is being adopted, symlinks are still being allowed.  This is only after verifying there is no information disclosure when symlinks are followed.  Although symlinks are not likely useful in many of these cases, this choice to still allow symlinks is to maintain compatibility with potentially unknown use cases.  Future use of this API will almost certainly always specify a `false` for allowing symlinks.

jentfoo committed:
Careful handling when loading files

This commit attempts to provide a common API so that the decision of when to follow symlinks is a conscious decision.
Because Teleport (particularly the agent) runs in a privilege context, there is risk that following symlinks may allow information disclosure.

After review of the cases covered in this commit (and some additional cases where this API was not a natural fit), this does not appear to be a broad problem.  This commit however does fix the one known flaw described in the issue https://github.com/gravitational/teleport-private/issues/1009

zmb3 commented on lib/cgroup/cgroup.go:
```diff
@@ -171,7 +172,7 @@ func (s *Service) Place(sessionID string, pid int) error {
 // within a cgroup.
 func readPids(path string) ([]string, error) {
-	f, err := os.Open(path)
+	f, err := utils.OpenFile(path, false)
```
It's not clear at the call-site what the extra boolean is for. What do you think about a separate func for when we don't want to follow symlinks? Something like `OpenFileNoSymlinks(path)` (then we could just use the standard `os.Open` when we want the default behavior.

jentfoo replied:
I suppose we could as long as the symlink is verbose in naming `OpenFileAllowingSymlinks`.  The goal in having it be a parameter is that the developer must make a choice in what parameter is provided.

Let me know what you think would be less likely to have a copy paste error, I think either could be fine as long as the name is verbose.

zmb3 replied:
> The goal in having it be a parameter is that the developer must make a choice in what parameter is provided.

Well, the developer doesn't have to make a choice if they don't know that we've created our own set of utilities that replace what the standard library offers. Most Go devs are not going to reach for these by default (especially because it's buried in a `utils` package with other misc and unrelated code).

I think it'd be better to use the standard library function when you want the default behavior of the standard library, and to use our overrides when you need something different.

jentfoo replied:
The goal is that "need something different" should be our default.  In general we should not allow the following of symlinks.  Do you have any thoughts on how we can make using this pattern a more default behavior?

codingllama replied:
+1 for "OpenFileNoSymlink" (or some other variation) instead of OpenFile(string, bool). Because OpenFile already exists is not the same as os.Open, I suppose we ought to keep both.

A way to get this through the codebase is to replace all usages, as at least when doing maintenance it should draw attention. Other than that, we could try and put together a linter of sorts.

jentfoo replied:
Updated the function naming.  I don't quite want to remove the ability to use `os.OpenFile` yet.  I am hopeful that this can still become the default when you need to open a read only file.

That said, we do need to keep in mind symlinks when writing too

timothyb89 commented on lib/utils/fs.go:
```diff
@@ -106,6 +107,9 @@ func OpenFile(path string) (*os.File, error) {
 	if err != nil {
 		return nil, trace.ConvertSystemError(err)
 	}
+	if !allowSymlink && fi.Mode().Type()&os.ModeSymlink != 0 {
+		return nil, trace.BadParameter("symlink not allowed")
+	}
```
A couple notes from things I learned trying to do this for `tbot`:
1. This check needs to use `os.Lstat()` to ever return `true`, since `os.Stat()` follows symlinks. We're just reading the destination file info here, which can never be a symlink, even if broken.
2. This one depends on our threat model, but even `os.Lstat()` follows symlinks, if they occur in any preceding path component. It's also a bit hairy to recursively check each path component since there could be cycles and other shenanigans.
   * This caveat applies to any solutions claiming to prevent symlink attacks using some variation of `O_NOFOLLOW`, since that flag only ever applies to the final pathname component.

We use [`openat2()` with `RESOLVE_NO_SYMLINKS`](https://man7.org/linux/man-pages/man2/openat2.2.html) in tbot to prevent this sort of file redirection attack: https://github.com/gravitational/teleport/blob/master/lib/tbot/botfs/fs_linux.go#L62

Unfortunately this comes with some pretty big caveats: `openat2()` isn't supported on older kernels, and is very frequently disabled on container engines and various enterprise Linuxes. And it's Linux-only :/. It's okay as an incremental security improvement where available, though - tbot checks for `openat2()` availability and falls back to insecure writes.

timothyb89 replied:
It also looks like we evaluate symlinks ahead of time in `NormalizePath()`, so we probably need to be careful with its symlink-following behavior too :slightly_frowning_face:

jentfoo replied:
Thank you @timothyb89, this is very helpful.  Let me look at this closer

jentfoo replied:
That said, I am surprised to see an OS compatibility issue with `Lstat`: https://github.com/gravitational/teleport/actions/runs/6150975451

Should this be os conditional?

jentfoo replied:
I see the issue, working on a fix now

timothyb89 replied:
I'm pretty sure that covers it. I've been feeding it inputs and haven't been able to break it, so that's something :slightly_smiling_face:

jentfoo replied:
Same!  I have been mentally weighing if I should build unit tests to cover these cases.  The downside is building the directory structure for testing is a fair bit of complexity.  Let me know your thoughts

jentfoo committed:
Apply PR feedback to rename OpenFile

jentfoo committed:
fs.go: Fix symlink evaluation

jentfoo committed:
fs.go: Fix for rebuilding absolute paths in symlink check

Tener commented on lib/utils/fs.go:
```diff
@@ -84,27 +85,59 @@ func IsDir(path string) bool {
+	} else {
+		isAbs := filepath.IsAbs(newPath)
+		components := strings.Split(newPath, string(os.PathSeparator))
+		for i := 1; i <= len(components); i++ {
+			subPath := filepath.Join(components[:i]...)
+			if isAbs {
+				subPath = string(os.PathSeparator) + subPath
+			}
+			fi, err = os.Lstat(subPath)
+			if err != nil {
+				return nil, trace.ConvertSystemError(err)
+			} else if fi.Mode().Type()&os.ModeSymlink != 0 {
+				return nil, trace.BadParameter("symlink not allowed in path: %v", subPath)
```
This code will be somewhat painful to test, but I do think we should test it anyway due to its complexity.

Tener replied:
We may want to add the original file path in the error; see my comment in tests for example why this could be useful.
```suggestion
				return nil, trace.BadParameter("failed to open file %v, symlink not allowed in path: %v", path, subPath)
```

Tener commented:
Independently, these two options are meaningful in this context:
- https://docs.kernel.org/admin-guide/sysctl/fs.html#protected-hardlinks
- https://docs.kernel.org/admin-guide/sysctl/fs.html#protected-symlinks

These are available in Linux kernel 3.6+, released 11 years ago, and enabled in _most_ contexts (but not all).

I don't think such an option is available on MacOS, though. It surely isn't the default.

jentfoo committed:
fs_test.go: Add symlink testing

Tener commented on lib/utils/fs_test.go:
```diff
@@ -26,6 +26,80 @@ import (
+	// dir/file-s -> dir/file
+	// circular-s -> circular-s
+	// broken-s -> nonexistent
+	rootDir, err := os.MkdirTemp("", "symlink-test")
+	require.NoError(t, err)
+	defer os.RemoveAll(rootDir)
```
There is a handy utility, which handles cleanup too:
```suggestion
	rootDir := t.TempDir()
```

Tener commented on lib/utils/fs_test.go:
```diff
@@ -26,6 +26,80 @@ import (
+	require.NoError(t, err)
+
+	// Tests below can not be done with t.Parallel due to the need for directory cleanup in defer
+	t.Run("directFileOpenAllowed", func(t *testing.T) {
+		f, err := openFile(filePath, false)
+		require.NoError(t, err)
+		defer f.Close()
+	})
```
This one actually fails on my machine, because `/var` is symlink to `/private/var`:

```
=== RUN   TestOpenFileSymlinks/directFileOpenAllowed
    fs_test.go:73: 
        	Error Trace:	/Users/tener/code/teleport/lib/utils/fs_test.go:73
        	Error:      	Received unexpected error:
        	            	symlink not allowed in path: /var
        	Test:       	TestOpenFileSymlinks/directFileOpenAllowed
    --- FAIL: TestOpenFileSymlinks/directFileOpenAllowed (0.00s)
```

I wonder if we can make the error more informative, i.e. include the information about the file we were trying to access:
```
=== RUN   TestOpenFileSymlinks/directFileOpenAllowed
    fs_test.go:77: 
        	Error Trace:	/Users/tener/code/teleport/lib/utils/fs_test.go:77
        	Error:      	Received unexpected error:
        	            	failed to open file /var/folders/c0/_6x4ddps6zn5js_4y31xlz380000gn/T/TestOpenFileSymlinks2354108266/001/dir/file, symlink not allowed in path: /var
        	Test:       	TestOpenFileSymlinks/directFileOpenAllowed
    --- FAIL: TestOpenFileSymlinks/directFileOpenAllowed (0.00s)
```

See the suggestion above. For this particular test, I think this works:

```suggestion
	t.Run("directFileOpenAllowed", func(t *testing.T) {
		filePath, err = filepath.EvalSymlinks(filePath)
		require.NoError(t, err)
		f, err := openFile(filePath, false)
		require.NoError(t, err)
		defer f.Close()
	})
```

Alternatively, you may want to apply this right next to declaration of `filePath` variable:

```golang
	filePath := filepath.Join(dirPath, "file")
	_, err = os.Create(filePath)
	require.NoError(t, err)
	filePath, err = filepath.EvalSymlinks(filePath)
	require.NoError(t, err)
```

Tener commented:
From the compatibility point of view, this is a breaking change. I don't think it will affect many configurations, but it is possible, in principle, that one of the paths in which we will now disallow symlinks contains those. For example, you can have `/home/%username%/` be a home dir for users, but symlinked to `/mnt/networkshare/developers/%username%`. For that reason, I'd add appropriate label as well as avoid backports to old releases.

jentfoo committed:
Apply suggestions from code review

Co-authored-by: Krzysztof Skrzętnicki <krzysztof.skrzetnicki@goteleport.com>

jentfoo committed:
fs_test.go: Fix build for err reference

codingllama commented on lib/utils/fs.go:
```diff
@@ -84,27 +85,59 @@ func IsDir(path string) bool {
+	return openFile(path, true)
+}
+
+// OpenFileNoSymlinks will open a file, ensuring it's an actual file and not a directory or symlink.
```
nits:

```suggestion
// OpenFileNoSymlinks opens a file, ensuring it's an actual file and not a directory or symlink.
```

codingllama commented on lib/utils/fs.go:
```diff
@@ -84,27 +85,59 @@ func IsDir(path string) bool {
+	return s, nil
+}
+
+// OpenFileAllowingSymlinks will open a file, if it's a symlink the file referenced by the link will be associated to
+// the returned os.File.  This will return an error if the file is not found or is a directory.
```
nit:

```suggestion
// OpenFileAllowingSymlinks opens a file, if it's a symlink the file referenced by the link will be associated to
// the returned os.File.  Returns an error if the file is not found or is a directory.
```

jentfoo replied:
Wording updated in a216f6d14848719cd061eb29b1ed657247eaee25

codingllama commented on lib/utils/fs.go:
```diff
@@ -84,27 +85,59 @@ func IsDir(path string) bool {
+	return s, nil
+}
+
+// OpenFileAllowingSymlinks will open a file, if it's a symlink the file referenced by the link will be associated to
+// the returned os.File.  This will return an error if the file is not found or is a directory.
```
> if it's a symlink the file referenced by the link will be associated to
> // the returned os.File

What do you mean here? Could we say it resolves the links instead?

codingllama commented on lib/utils/fs.go:
```diff
@@ -84,27 +85,59 @@ func IsDir(path string) bool {
-// OpenFile opens  file and returns file handle
-func OpenFile(path string) (*os.File, error) {
-	newPath, err := NormalizePath(path)
+// openFile opens a file and returns file handle. In general symlinks should not be allowed to reduce the risk of
+// privilege escalation from Teleports elevated privileges to potentially less privileged users accidentally.
```
I think we can kill this comment, the public methods should cover all of this.

```suggestion
```

jentfoo replied:
Removed in a216f6d14848719cd061eb29b1ed657247eaee25

codingllama commented on lib/utils/fs.go:
```diff
@@ -84,27 +85,59 @@ func IsDir(path string) bool {
+			if isAbs {
+				subPath = string(os.PathSeparator) + subPath
+			}
+			fi, err = os.Lstat(subPath)
```
Is the Lstat skip in the first path component on purpose?

jentfoo replied:
I don't believe there is any skipping of components, can you explain further?

codingllama replied:
The for starts with `i := 1`, so for relative paths we are not checking if the very first component is a symlink, unless I read it wrong?

jentfoo replied:
The components are selected using the sub-slice `[:i]`.  So the beginning part of the components will always be included in every iteration.  We start at `1` because the sub-slice is exclusive to the index.  This is also why the conditional is `<=`.

To me this seemed like the most obvious implementation but now I am wondering your recommendation to rebuild the path using the range of the components is easier to read.  Let me know your thoughts.

codingllama replied:
Gotcha, I did read it wrong. Thanks for explaining.

codingllama commented on lib/utils/fs.go:
```diff
@@ -84,27 +85,59 @@ func IsDir(path string) bool {
+			return nil, trace.ConvertSystemError(err)
+		}
+	} else {
+		isAbs := filepath.IsAbs(newPath)
+		components := strings.Split(newPath, string(os.PathSeparator))
+		for i := 1; i <= len(components); i++ {
+			subPath := filepath.Join(components[:i]...)
+			if isAbs {
+				subPath = string(os.PathSeparator) + subPath
+			}
```
Suggestion: Use a for-range, Join addictively, check first path component, spacing.

```suggestion
		components := strings.Split(newPath, string(os.PathSeparator))
		var subpath string
		for _, p := range components {
			subpath = filepath.Join(subpath, p)

			// The first component in an absolute path is empty.
			if subpath == "" {
				subPath = string(os.PathSeparator)
			}

```

jentfoo replied:
It's unclear to me why this is better, let me know your thoughts

codingllama replied:
for-ranges are usually preferable as there is no need to deal with `i`. This ties in with a previous comment.

The snippet also replaces filepath.IsAbs with an empty string check, which I'd expect to be much cheaper - not a big deal here, but why not do it?

Other than those, it's a toss, hence the "suggestion" tag.

jentfoo replied:
Updated, you convinced me this is more readable.  Thank you!

codingllama replied:
Thanks!

codingllama commented on lib/utils/fs.go:
```diff
@@ -84,27 +85,59 @@ func IsDir(path string) bool {
+			if err != nil {
+				return nil, trace.ConvertSystemError(err)
+			} else if fi.Mode().Type()&os.ModeSymlink != 0 {
+				return nil, trace.BadParameter("failed to open file %v, symlink not allowed in path: %v", path, subPath)
```
nit: Skip filler words like "failed" in error chains.

```suggestion
				return nil, trace.BadParameter("opening file %v, symlink not allowed in path: %v", path, subPath)
```

codingllama commented on lib/utils/fs_test.go:
```diff
@@ -26,6 +26,80 @@ import (
+	err = os.Symlink(filepath.Join(rootDir, "nonexistent"), brokenSymlink)
+	require.NoError(t, err)
+
+	// Tests below can not be done with t.Parallel due to the need for directory cleanup in defer
```
Let's refactor this into a table-driven test, it's a good fit.

https://github.com/golang/go/wiki/TableDrivenTests

codingllama commented on lib/utils/fs_test.go:
```diff
@@ -26,6 +26,80 @@ import (
+	t.Run("directFileOpenAllowed", func(t *testing.T) {
+		filePath, err = filepath.EvalSymlinks(filePath)
+		require.NoError(t, err)
+		f, err := openFile(filePath, false)
```
Please test the public functions instead.

jentfoo replied:
Now that the public functions include OS specific logic, I don't believe that would be a good fit for testing.

codingllama replied:
If that's a concern, then we should be writing per-OS testing. I would still encourage you to test the public API and not internals.

jentfoo replied:
As in have the testing define the expected state depending on the OS it's run on?  Or have individual tests for each OS?

The first seems pretty reasonable if you would like me to update to that.  I will probably add it as additional testing rather than replace these.

codingllama replied:
I meant the latter - we have a vanilla test suite for the "regular" scenarios, but test macOS vs Linux behavior separately. We have a few packages that behave like that, for example webauthn tends to vary between darwin, linux and windows.

jentfoo replied:
I am not sure I fully agree.  Because of the need to have a more complex setup that interacts with the filesystem I think it's best if this test is as centrally located as possible.

That said I did add testing around the public functions and the OS specific behaviors with hard links.  Let me know what you think.  Thank you!

codingllama commented on lib/utils/fs_test.go:
```diff
@@ -26,6 +26,80 @@ import (
+	err := os.Mkdir(dirPath, 0755)
+	require.NoError(t, err)
+
+	filePath := filepath.Join(dirPath, "file")
```
```suggestion
	dirFilePath := filepath.Join(dirPath, "file")
```

So the hierarchy is clearer from the names? Same for others.

codingllama commented on lib/utils/fs_test.go:
```diff
@@ -26,6 +26,80 @@ import (
+	})
+	t.Run("symlinkFileOpenDenied", func(t *testing.T) {
+		_, err := openFile(symlinkToFile, false)
+		require.Error(t, err)
```
Maybe require.ErrorContains, so we can be sure we are hitting the expected condition?

```suggestion
		require.ErrorContains(t, err, "symlink")
```

Same for others.

(In a minor nit, we could be using `assert` instead of `require` here, which is the better habit.)

jentfoo replied:
Updated with the other test updates in 734437505efc1fb43e6a9a478b2d2516a077085d

codingllama commented:
> From the compatibility point of view, this is a breaking change

Could we mention in the changelog, for example, the places where symlinks are not allowed anymore? (Ie, the places affected by the change.)

@jentfoo

jentfoo commented:
Compatibility options are being discussed here: https://github.com/gravitational/teleport-private/issues/1009#issuecomment-1716028879

I think we can find a way to backport this improvement while maintaining compatibility.

codingllama commented:
> I think we can find a way to backport this improvement while maintaining compatibility.

I fail to see how the behavioral change would be non-breaking. I'm not saying to not backport, no strong opinions there on my part, just asking that we document whatever gets changed.

jentfoo commented:
The compatibility solution was not as easy as it appeared.  For that reason I have removed the backport labels, but would like to discuss if we can get this into Teleport 14.

@codingllama and @Tener I would appreciate another review, thank you for all the feedback so far!

jentfoo committed:
fs.go: Apply PR feedback and consider Hardlinks

After PR discussion it was highlighted that MacOS does not guard against hardlinks in the same way linux does.  For that reason this implementation has been updated with OS conditional logic to validate against hardlinks.

jentfoo committed:
fs.go: Switch loop to range over components instead of index

codingllama commented on lib/utils/fs_test.go:
```diff
@@ -26,6 +26,144 @@ import (
+			f, err := openFile(tt.filePath, tt.allowSymln, tt.allowHardln)
+			if tt.expectErr == "" {
+				require.NoError(t, err)
+				f.Close()
```
nit: Close() should happen based on the actual outcome of the method, not be based on the test expectations, as the latter may break.

codingllama reviewed (commented):
Looks good, holding out just in case you do more test changes.

jentfoo committed:
Minor improvements from PR feedback

jentfoo committed:
fs_test.go: Test public OpenFile API's and include OS specific validation

r0mant commented on lib/utils/fs.go:
```diff
@@ -84,30 +87,69 @@ func IsDir(path string) bool {
+	}
+	if !allowMultipleHardlinks {
+		// hardlinks can only exist at the end file, not for directories within the path
+		if statT, ok := fi.Sys().(*syscall.Stat_t); ok && statT.Nlink > 1 {
```
This will not compile on Windows, but I don't remember off the top of my head if `tsh` (the only tool we build for Windows right now) pulls in this package. Might be worth checking before trying to merge this.

r0mant replied:
And yeah, `tctl` as well as @zmb3 mentioned below. Probably worth it to split this out into OS-specific files via build tags.

r0mant reviewed (approved):
I don't have objections to including this change in Teleport 14 (bar potential compilation issue) but @jentfoo can you please add a section describing the user impact of this change to our [draft release notes](https://docs.google.com/document/d/1NRaue1huhkzwmbDM6it8cNgaO9rZQ2b9O0w0nrG5W5Y/edit) under the "Breaking changes".

I also assume you did a smoke test with this change (e.g. starting teleport and starting a session)?

zmb3 commented:
We also build tctl for windows. We don't ship it yet, but we do build it on push and a lot of work went in to getting this far so please don't break it 🙂

jentfoo committed:
Fix windows build

Make hardlink count lookup code build conditional to avoid undefined syscall.Stat_t.

codingllama commented on lib/utils/fs_unix.go:
```diff
@@ -19,7 +19,20 @@ limitations under the License.
 	return path
 }
+
+func getHardLinkCount(fi os.FileInfo) (bool, uint64) {
```
nit: It's more usual to place the boolean return last:

```suggestion
func getHardLinkCount(fi os.FileInfo) (uint64, bool) {
```

In this case, I suppose returning just `0` or `-1` (without a bool) could work too. Up to you.

jentfoo replied:
I changed the order.  The original order made sense to me based off how they should be evaluated, but no preference really.

I think it makes sense to return a bool for two reasons:
* I want to return a uint to match the `Stat_T` typing and avoid overflow risks
* Technically having a zero hardlink count is a valid condition if the file handle does not exist on disk (for example an open file handle that has been deleted)

codingllama replied:
I'm happy with the changed order, thanks.

jentfoo committed:
utils.getHardLinkCount result order update from PR feedback

zmb3 commented on lib/utils/fs_unix.go:
```diff
@@ -19,7 +19,20 @@ limitations under the License.
+		return true, statT.Nlink
+	} else {
+		return false, 0
+	}
```
I'd write this slightly differently in order to eliminate the else and align the happy path to the left.

```
statT, ok := fi.Sys().(*syscall.Stat_t)
if !ok {
    return false, 0
}
return true, statT.Nlink
```

jentfoo replied:
This seems basically the same to me.  I tend to prefer avoiding `!` conditionals when logic for both states needs to be handled anyways.

zmb3 commented on lib/utils/fs_windows.go:
```diff
@@ -27,7 +27,10 @@ limitations under the License.
 // was causing flock.Flock.TryRLock to return either "access denied" or "The process cannot access
 // the file because it is being used by another process".
```
This comment used to apply to the entire file, not is only applies to `getPlatformLockFilePath`. I'd move it down.

jentfoo replied:
Moved

zmb3 commented on lib/utils/fs_windows.go:
```diff
@@ -39,3 +42,7 @@ func getPlatformLockFilePath(path string) string {
 }
+
+func getHardLinkCount(fi os.FileInfo) (bool, uint64) {
+	return false, 0
```
Maybe add a comment

```
// hard links not supported for Windows
```

jentfoo replied:
Added a comment describing the current situation

jentfoo committed:
fs_windows.go: Comment improvements from PR feedback

jentfoo commented:
@r0mant I have a suggestion in the release notes for how to describe this.  I tried to describe it fairly generically, but the only known use case to me is the loading of the `.tsh/environment` file.  Feel free to DM me if you think there is any better way to describe this for our customers.

Also I have added the `branch/v14` backport label so we can include this as part of Teleport 14

jentfoo committed:
fs: Fix build for OSX

jentfoo committed:
Disable lint error due to unecessary cast on linux
